### PR TITLE
Contact Shadows

### PR DESCRIPTION
**Pull request by JulienIgnace-Unity**
First version of Contact Shadows for directional lights.
Enabled through the Light Editor.

### CODE REVIEW

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs [hunk 1]
```diff
@@ -1971,23 +1973,41 @@ public struct LightingPassOptions
 
         public void RenderDeferredDirectionalShadow(HDCamera hdCamera, RenderTargetIdentifier deferredShadowRT, RenderTargetIdentifier depthTexture, CommandBuffer cmd)
         {
-            if (m_CurrentSunLight == null)
+            if (m_CurrentSunLight == null || m_CurrentSunLight.GetComponent<AdditionalShadowData>() == null)
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs]: ?

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/RenderPipeline/FrameSettings.cs [hunk 2]
```diff
@@ -40,6 +41,7 @@ public class FrameSettings
         // Lighting
         // Setup by users
         public bool enableShadow = true;
+        public bool enableContactShadows = true;
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/RenderPipeline/FrameSettings.cs]: should be false by default ?

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs (hunk 1 above)
Comment (sebastienlagarde) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs]: if no additionalshadow data, just setup false in 
bool enableContactShadows = m_FrameSettings.enableContactShadows && asd.enableContactShadows && asd.contactShadowLength >

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/DeferredDirectionalShadow.compute [hunk 3]
```diff
@@ -1,28 +1,128 @@
...
+	UpdatePositionInput(depth, UNITY_MATRIX_I_VP, UNITY_MATRIX_VP, posInput);
 
-    float depth = LOAD_TEXTURE2D(_MainDepthTexture, posInput.positionSS).x;
-    UpdatePositionInput(depth, UNITY_MATRIX_I_VP, UNITY_MATRIX_VP, posInput);
+	ShadowContext shadowContext = InitShadowContext();
+	float shadow = GetDirectionalShadowAttenuation(shadowContext, posInput.positionWS, float3(0.0, 0.0, 0.0), (uint)_DirectionalShadowIndex, float3(0.0, 0.0, 0.0));
 
-    ShadowContext shadowContext = InitShadowContext();
-    float shadow = GetDirectionalShadowAttenuation(shadowContext, posInput.positionWS, float3(0.0, 0.0, 0.0), (uint)_DirectionalShadowIndex, float3(0.0, 0.0, 0.0));
+#if ENABLE_CONTACT_SHADOWS
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/DeferredDirectionalShadow.compute]: ifdef ?

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Editor/Lighting/HDLightEditor.cs [hunk 4]
```diff
@@ -182,6 +195,7 @@ void DrawFeatures()
             {
                 bool shadowsEnabled = EditorGUILayout.Toggle(CoreEditorUtils.GetContent("Enable Shadows"), settings.shadowsType.enumValueIndex != 0);
                 settings.shadowsType.enumValueIndex = shadowsEnabled ? (int)LightShadows.Hard : (int)LightShadows.None;
+                EditorGUILayout.PropertyField(m_AdditionalShadowData.enableContactShadows, CoreEditorUtils.GetContent("Enable Contact Shadows"));
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Editor/Lighting/HDLightEditor.cs]: should be only on directional light

Comment (sebastienlagarde) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Editor/Lighting/HDLightEditor.cs]: also add a comment somewhere that only the dir ligth with (cascade) shadow can enable shadow contact

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs (hunk 1 above)
Comment (JulienIgnace-Unity) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/LightLoop/LightLoop.cs]: For some reason, sometime in the editor, additionalShadowData is null :/
I suspect something with preview.

**Pull request merged**